src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyService.java

```diff
@@ -279,6 +287,59 @@ public void deleteCompetency(Competency competency, Course course) {
         competencyRepository.deleteById(competency.getId());
     }
 
+    /**
+     * Finds a competency by its id and fetches its lecture units, exercises and progress for the provided user. It also fetches the lecture unit progress for the same user.
+     * <p>
+     * As Spring Boot 3 doesn't support conditional JOIN FETCH statements, we have to retrieve the data manually.
+     *
+     * @param competencyId The id of the competency to find
+     * @param userId       The id of the user for which to fetch the progress
+     * @return The found competency
+     */
+    public Competency findCompetencyWithExercisesAndLectureUnitsAndProgressForUser(Long competencyId, Long userId) {
```

Comment:
we should also double check if this query can become slow

Reply:
It always has three database calls. The runtime of the for loops then dependy linearly on the number of lecture units.

While it's obviously less efficient than retrieving all at once, I don't see any reason of concern. Or do you insist on a timed test?

Reply:
I just had time and did it anyway:

All results
- are given in seconds
- contain 1000 calls to the endpoint calling this method/the previous repository call
- are part of a `RepeatedTest(10)`

| # | Δ dev  | Δ new  |
| :-| :-------: | :------: |
| 0 | 21.511 | 17.258 |
| 1 | 10.922 | 09.560 |
| 2 | 09.414 | 09.162 |
| 3 | 08.596 | 08.024 |
| 4 | 08.108 | 07.623 |
| 5 | 08.196 | 07.410 |
| 6 | 08.228 | 07.020 |
| 7 | 07.531 | 07.131 |
| 8 | 08.218 | 07.135 |
| 9 | 07.660 | 07.818 |

In conclusion: My computer is not a dedicated testing environment, but I think we can safely assume that we do not have a serious performance loss through these changes.